README.md
# Wifi For Pi

The purpose of this module is to provide tools to connect to a Wifi

# Table of Contents

- [Installation](#installation) - How to install the module
- [API](#api) - Commands available
  - [check(ssid, callback)](#checkssid-callback) - State of the network with the specified ssid
  - [connectTo(details, callback)](#connecttodetails-callback) - State of the network with the specified ssid
  - [connectToId(networkId, callback)](#connecttoidnetworkid-callback) - Connects to a network with the parameters specified (This can connect to open and secure networks including 802.1x)
  - [connect(ssid, password, callback)](#connectssid-password-callback) - Connects to a network with the ssid specified using the password provided
  - [connectOpen(ssid, callback)](#connectopenssid-callback) - Connects to an open network with the ssid specified
  - [connectEAP(ssid, password, callback)](#connecteapssid-password-callback) - Connects to a network with the ssid specified using the password provided
  - [disconnect(callback)](#disconnectcallback) - Disconnects from the network on the current interface
  - [detectSupplicant(callback)](#detectsupplicantcallback) - Looks for a running wpa_supplicant process and if so returns the config file and interface used
  - [interfaceDown(callback)](#interfacedowncallback) - Drops the interface provided
  - [interfaceUp(callback)](#interfaceupcallback) - Raises the interface provided
  - [killSupplicant(callback)](#killsupplicantcallback) - Kills the supplicant process for the specified interface
  - [listNetworks(callback)](#listnetworkscallback) - List the networks in an array, each network has Network ID, SSID, BSSID and FLAGS
  - [restartInterface(callback)](#restartinterfacecallback) - Restarts the interface provided
  - [scan(callback)](#scancallback) - Scan available wifi networks
  - [setCurrentInterface(iface)](#setcurrentinterfaceiface) - Specify the interface to use
  - [startSupplicant(callback)](#startsupplicantcallback) - Starts a wpa_supplicant instance
  - [status(iface, callback)](#statusiface-callback) - Show status parameters of the interface specified, if no interface is provided the selected one is used
- [Notes](#notes)


# Installation

On your project root execute:

``` shell
$ npm install --save pi-wifi
```

# API

The following is the list of functions available:


## check(ssid, callback)
The **check** function is used to return the state of the network with the specified ssid

``` javascript
var piWifi = require('pi-wifi');

piWifi.check('myTestNetwork', function(err, result) {
  if (err) {
    return console.error(err.message);
  }
  console.log(result);
});

// => If everything is working
// { selected: true, connected: true, ip: '192.168.0.12' }
// => Alternatively
// { selected: false, connected: false }
```

## connectTo(details, callback)
The **connectTo** function is used to connect to a network with the parameters specified (This can connect to open and secure networks including 802.1x)

The **details** object can contain the following:

- ***key_mgmt*** You can specify the type of security to use. (Optional)
- ***ssid*** (required for secure and enterprise networks)
- ***username*** (required for enterprise networks)
- ***password*** (required for enterprise networks)
- ***eap***
- ***phase1***
- ***phase2***

``` javascript
var piWifi = require('pi-wifi');

var networkDetails = {
  ssid: 'MyNetwork',
  username: 'demo',
  password: 'swordfish'
};

//A simple connection
piWifi.connectTo(networkDetails, function(err) {
  if(!err) {
    console.log('Network created successfully!');
  } else {
    console.log(err.message); //Failed to connect
  }
});

//After creating a succesful connection, you could use the function check to verify
var ssid = 'MyOpenNetwork';
piWifi.connectTo({ssid: ssid}}, function(err) {
  if (!err) { //Network created correctly
    setTimeout(function () {
      piwifi.check(ssid, function (err, status) {
        if (!err && status.connected) {
          console.log('Connected to the network ' + ssid + '!');
        } else {
          console.log('Unable to connect to the network ' + ssid + '!');
        }
      });
    }, 2000);
  } else {
    console.log('Unable to create the network ' + ssid + '.');
  }
});

```

## connectToId(networkId, callback)
The **connectToId** function is used enables the network, saves the configuration and then select the network with the id provided

``` javascript
var piWifi = require('pi-wifi');

piWifi.connectToId(2, function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Successful connection!');
});
```

## connect(ssid, password, callback)
The **connect** function is used to connect to a network with the ssid specified using the password provided
***(Avoid using this as it is only for backwards compatibility and may be deprecated in the future)***

``` javascript
var piWifi = require('pi-wifi');

piWifi.connect('myTestNetwork', 'MyTestPassword', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Successful connection!');
});
```

## connectOpen(ssid, callback)
The **connectOpen** function is used to connect to an open network with the ssid specified
***(Avoid using this as it is only for backwards compatibility and may be deprecated in the future)***

``` javascript
var piWifi = require('pi-wifi');

piWifi.connectOpen('myTestNetwork', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Successful connection!');
});
```


## connectEAP(ssid, password, callback)
The **connectEAP** function is used to connect to a network with the ssid specified using the password provided
***(Avoid using this as it is only here for consistency and may be deprecated in the future)***

``` javascript
var piWifi = require('pi-wifi');

piWifi.connectEAP('myTestNetwork', 'MyTestUsername', 'MyTestPassword', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Successful connection!');
});
```

## disconnect(callback)
The **disconnect** function is used to disconnect from the network on the current interface
``` javascript
var piWifi = require('pi-wifi');

piWifi.disconnect(function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Disconnected from network!');
});
```

## detectSupplicant(callback)
The **detectSupplicant** function is used to look for a running wpa_supplicant process and if found returns the config file and interface used

``` javascript
var piWifi = require('pi-wifi');

piWifi.detectSupplicant(function(err, iface, configFile) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Supplicant running in interface', iface, 'using the configuration file', configFile);
});
```

## interfaceDown(callback)
The **interfaceDown** function is used to drop the interface provided

``` javascript
var piWifi = require('pi-wifi');

piWifi.interfaceDown('wlan0', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Interface dropped succesfully!');
});
```

## interfaceUp(callback)
The **interfaceUp** function is used to raise the interface provided

``` javascript
var piWifi = require('pi-wifi');

piWifi.interfaceUp('wlan0', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Interface raised succesfully!');
});
```

## killSupplicant(callback)
The **killSupplicant** function is used to kill the supplicant process for the specified interface
``` javascript
var piWifi = require('pi-wifi');

piWifi.killSupplicant('wlan0', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Supplicant process terminated!');
});
```

## listNetworks(callback)
The **listNetworks** function is used to list the networks in an array, each network has Network ID, SSID, BSSID and FLAGS

``` javascript
var piWifi = require('pi-wifi');

piWifi.listNetworks(function(err, networksArray) {
  if (err) {
    return console.error(err.message);
  }
  console.log(networksArray);
});

// =>
// [{ network_id: 0, ssid: 'MyNetwork', bssid: 'any', flags: '[DISABLED]' },
// { network_id: 1, ssid: 'Skynet', bssid: 'any', flags: '[CURRENT]' }]
```

## restartInterface(callback)
The **restartInterface** function is used to restart the interface provided

``` javascript
var piWifi = require('pi-wifi');

piWifi.restartInterface('wlan0', function(err) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Interface restarted succesfully!');
});
```

## scan(callback)
The **scan** function is used to scan available wifi networks

``` javascript
var piWifi = require('pi-wifi');

piWifi.scan(function(err, networks) {
  if (err) {
    return console.error(err.message);
  }
  console.log(networks);
});

// =>
//[ 
//  { bssid: 'aa:bb:cc:dd:ee:ff',
//    frequency: 2462,
//    signalLevel: -40,
//    flags: '[WPA2-PSK-CCMP][WPS][ESS]',
//    ssid: 'MyNetwork' },
//  { bssid: '11:22:33:44:55:66',
//    frequency: 2462,
//    signalLevel: -28,
//    flags: '[WPA2-PSK-CCMP][ESS]',
//    ssid: 'AnotherNetwork' },
//  { bssid: 'aa:11:bb:22:cc:33',
//    frequency: 2462,
//    signalLevel: -33,
//    flags: '[WPA2-EAP-CCMP-preauth][WPS][ESS]',
//    ssid: 'MyEnterpriseNetwork' },
//  { bssid: 'c0:56:27:44:3b:9c',
//    frequency: 2412,
//    signalLevel: -59,
//    flags: '[WPA-PSK-CCMP+TKIP][WPA2-PSK-CCMP+TKIP][ESS]',
//    ssid: 'MyGuestsNetwork' 
//  }
//]
```

## setCurrentInterface(iface)
The **setCurrentInterface** function is used to specify the interface to use as default

``` javascript
var piWifi = require('pi-wifi');
piWifi.setCurrentInterface('wlan1');

piWifi.status('wlan0', function(err, status) {
  if (err) {
    return console.error(err.message);
  }
  console.log(status);
});

```

## startSupplicant(callback)
The **startSupplicant** function is used to start a wpa_supplicant instance

``` javascript
var piWifi = require('pi-wifi');

piWifi.startSupplicant({iface: 'wlan0', config: '/etc/wpa_supplicant/wpa_supplicant.conf', dns: '/etc/resolv.conf'}, function(err, networks) {
  if (err) {
    return console.error(err.message);
  }
  console.log('Supplicant instance successfully started!');
});

```

## status(iface, callback)
The **status** function is used to show status parameters of the interface specified, if no interface is provided the selected one is used

``` javascript
var piWifi = require('pi-wifi');

piWifi.status('wlan0', function(err, status) {
  if (err) {
    return console.error(err.message);
  }
  console.log(status);
});

// =>
//{
//  bssid: '2c:f5:d3:02:ea:d9',
//  frequency: 2412,
//  mode: 'station',
//  key_mgmt: 'wpa2-psk',
//  ssid: 'MyNetwork',
//  pairwise_cipher: 'CCMP',
//  group_cipher: 'CCMP',
//  p2p_device_address: 'aa:bb:cc:dd:ee:ff',
//  wpa_state: 'COMPLETED',
//  ip: '10.20.30.40',
//  mac: 'a1:b2:c3:d4:e5:f6',
//  uuid: 'e1cda789-8c88-53e8-ffff-31c304580c22',
//  id: 0
//}
```

## Notes
Tested on Raspberry Pi 3


HISTORY.md
## [1.2.0]
### Added
- Readme
   - connectTo (A global connection function instead of one per security type)
   -   connectToId (In case you know already know the networkId)
   -    connect (Keeping this for backwards compatibility) DEPRECATED?
   -    connectOpen (Keeping this for backwards compatibility) DEPRECATED?

   -   connectEAP (Adding it for consistency...) DEPRECATED?
   -   disconnect
   -   detectSupplicant
   -   interfaceDown: interfaceDown,
   -   interfaceUp: interfaceUp,
   -  killSupplicant: disableSupplicant,
   -  listNetworks (Lists already known networks)
   


index.js
/* Intended to be run ON the Raspberry Pi */

var tools = require('wireless-tools');
var async = require('async');

const defaultOpenKey = 'NONE';
const defaultEnterpriseKey = 'WPA-EAP';
const defaultSupplicantConfigFile = '/etc/wpa_supplicant/wpa_supplicant.conf';
const defaultDNSFile = '/etc/resolv.conf';
const defaultInterface = 'wlan0';

var currentInterface = defaultInterface;

const commands = {
  detectSupplicant: 'ps -fea | grep -v grep | grep wpa_supplicant',
  interfaceDown: 'ifdown --force :INTERFACE',
  interfaceUp: 'ifup :INTERFACE',
  scan: 'sudo iwlist wlan0 scan | grep ESSID | cut \'"\' -f2',
  startSupplicant: 'sudo wpa_supplicant -Dwext -c :CONFIG_FILE -B -i :INTERFACE  && sudo chattr -i :DNS_FILE',
  wpaDisconnect: 'wpa_cli disconnect',
  wpaInterface: 'wpa_cli interface :INTERFACE',
  wpaList: 'wpa_cli list_networks'
};


var exec = require('child_process').exec;

/**
* @method scan
* @description scan available wifi networks
* @param {Function} callback
* @return {Array[Object]}
*/
function scan(callback) {
  tools.wpa.scan(currentInterface, function (err, data) {
    if (err || !data.hasOwnProperty('result') || data.result !== 'OK') return callback(err, data);
    setTimeout(function () {
      //console.log('Scan results:');
      tools.wpa.scan_results(currentInterface, function (err, networks) {
        callback(err, networks);
      });
    }, 1000);
  });
}


/**
* @method findConnection
* @description Search a network on the known networks list and return it's network id, fails if not found
* @param {string} ssid Network ssid
* @param {Function(err, networkId)} callback Returns error if the network isn't found, id of the network found
*/
function findConnection(ssid, callback) {

  var networkId;
  //console.log('Listing...');
  listNetworks(function (err, networksArray) {
    if (!err) {
      //console.log('Looking for the network...');
      for (var i = 0; i < networksArray.length; i++) {
        if (networksArray[i].ssid === ssid) {
          //console.log('Network ' + networksArray[i].network_id + ' found');
          return callback(err, networksArray[i].network_id);
        }
      }
    }
    //console.log('Network not found...');
    return callback(err, networkId);
  });
}

/**
* @method setCurrentInterface
* @description Specify the interface to use
* @param {string} iface Intertface to set
* @param {function} callback Returns error if unable to set the interface
*/
function setCurrentInterface(iface, callback) {
  exec(replaceInCommand(commands.wpaInterface, { interface: iface }), function (err) { 
    if (!err) currentInterface = iface;
    return callback(err);
  });
}

/**
* @method checkConnectionDetails
* @description Verifies the parameters details and if it isn't properly formed it will return an error, undefined otherwise
* @param {Object} params Json object with the network parameters
* @param {Function(err)} callback Returns error if the connection isn't successful
*/
function checkConnectionDetails(/*params*/) {
  var err;
  //TODO implement this verification (e.g.: If a secure network is missing a password error out)
  return err;
}

/**
 * @method prepareConnectionDetails
 * @param {Object} details
 * @return {Object} Returns an adjusted json object with the proper format and default parameters required
 */
function prepareConnectionDetails(details) {

  var params = {};
  if (details.hasOwnProperty('ssid')) params.ssid = '\'"' + details.ssid + '"\'';  //Add ssid

  if (!details.hasOwnProperty('password')) {
    if (!details.hasOwnProperty('key_mgmt')) params.key_mgmt = defaultOpenKey; //Set key_mgmt for an open network
  } else if (!details.hasOwnProperty('username')) { //If no user then it is a regular secured network
    params.psk = '\'"' + details.password + '"\'';
  } else { //If it has password and user it must be an enterprise network
    params.key_mgmt = defaultEnterpriseKey; //Set key_mgmt for default enterprise network (802.1x)
    params.identity = '\'"' + details.username + '"\'';
    params.password = '\'"' + details.password + '"\'';
    if (!details.hasOwnProperty('eap')) params.eap = 'PEAP';
    //if (!details.hasOwnProperty('phase2')) params.phase2 = '\'"autheap=MSCHAPV2"\'';
  }

  if (details.hasOwnProperty('eap')) params.eap = details.eap;
  if (details.hasOwnProperty('phase1')) params.phase1 = details.phase2;
  if (details.hasOwnProperty('phase2')) params.phase2 = details.phase2;
  if (details.hasOwnProperty('key_mgmt')) params.key_mgmt = details.key_mgmt; //Add Key management if found

  return params;
}

/**
* @method connection
* @description Connects to a network with the parameters specified (This can connect to open and secure networks including EAP 802.1x)
* @param {Object} details Network details
* - {string} key_mgmt You can specify the type of security to use. (Optional)
* - {string} ssid (Optional, required for secure and enterprise networks)
* - {string} username (Optional, required for enterprise networks)
* - {string} password (Optional, required for secure or enterprise networks)
* @param {Function(err)} callback Returns error if the network creation isn't successful
*/
function connection(details, callback) {
  var netId;

  var err = checkConnectionDetails(details);
  if (err) return callback(err);

  var params = prepareConnectionDetails(details);

  findConnection(details.ssid, function (err, networkId) { //Look for an existing connection with that SSID
    async.series({
      remove: function(next) { //Remove the network if found
        if (networkId === undefined) return next(undefined);
        tools.wpa.remove_network(currentInterface, networkId, next);
      },
      create: function(next) { //Create a new network
        createConnection(function (err, networkId) {
          if (!err) netId = networkId;
          //console.log('Created network:', netId, err);
          next(err);
        });
      },
      setup: function (next) { //Add provided parameters to network
        setupConnection(netId, params, next);
      }, //Can't use async.apply due to netId being shared
      start: function(next) { connectToId(netId, next); } //Actually connect to the network
    }, function (err, results) {
      if (err && results.create) tools.wpa.remove_network(currentInterface, netId, callback);
      else callback(err);
    });
  });
}


/**
* @method openConnection
* @description Connects to an open network with the ssid specified
* @param {string} ssid Network ssid
* @param {Function(err)} callback Returns error if the connection isn't successful
*/
function openConnection(ssid, callback) {

  var params = {
    ssid: ssid
  };
  connection(params, callback);
}


/**
* @method enterpriseConnection
* @description Connects to a network with the ssid specified using the password provided
* @param {string} ssid Network ssid
* @param {string} username User/identity to use on authentication
* @param {string} password Password to use on authentication
* @param {Function(err)} callback Returns error if the connection isn't successful
*/
function enterpriseConnection(ssid, username, password, callback) {

  var params = {
    ssid: ssid,
    username: username,
    password: password
  };
  connection(params, callback);
}


/**
* @method secureConnection
* @description Connects to a network with the ssid specified using the password provided
* @param {string} ssid Network ssid
* @param {string} ssid Network psk
* @param {Function(err)} callback Returns error if the connection isn't successful
*/
function secureConnection(ssid, password, callback) {

  var params = {
    ssid: ssid,
    password: password
  };
  connection(params, callback);

}


/** 
* @method createConnection
* @description Creates a connection record and returns its network id if successful
* @param {Function(err, networkId)} callback Returns error if the network creation fails, Network id
*/
function createConnection(callback) {
  tools.wpa.add_network(currentInterface, function (err, netId) {
    if (!err && netId.hasOwnProperty('result')) netId = netId.result;
    callback(err, netId);
  });
}


/**
* @method setupConnection
* @description Sets the parameters for a network
* @param {Integer} networkId Network id
* @param {Object} params Json object with network parameters to set (Name, Value)
* @param {Function(err)} callback Returns error if the process fails
*/
function setupConnection(netId, params, callback) {
  async.eachOf(params, function (value, key, next) { //For each one of the parameters listed
    console.log('>>> Setting', key, 'with', value);
    setNetworkParameter(currentInterface, netId, key, value, next); //Set it to the network
  }, callback);
}


/**
 * @method setNetworkParameter
 * @description Sets a network parameter
 * @param {string} interface Interface to use
 * @param {integer} networkId Network to set it to
 * @param {string} name Parameter key/name
 * @param {string} value Parameter value
 * @param {function} callback Returns an error if the parameter wasn't set
 */
function setNetworkParameter(interface, networkId, name, value, callback) {
  tools.wpa.set_network(interface, networkId, name, value, callback);
}


/**
* @method status
* @description Show status parameters of the interface specified, if no interface is provided the selected one is used
* @param {string} iface Interface to get status from. (If not provided it defaults to the currently selected one)
* @param {Function(err)} callback Returns error if the process fails, status JSON object with the interface status
* @example
*
* status('wlan0', function(err, status){
*   if(!err) console.log(status);
* });
* // => 
* {
*   bssid: '2c:f5:d3:02:ea:d9',
*   frequency: 2412,
*   mode: 'station',
*   key_mgmt: 'wpa2-psk',
*   ssid: 'Fake-Wifi',
*   pairwise_cipher: 'CCMP',
*   group_cipher: 'CCMP',
*   p2p_device_address: 'e4:28:9c:a8:53:72',
*   wpa_state: 'COMPLETED',
*   ip: '10.34.141.168',
*   mac: 'e4:28:9c:a8:53:72',
*   uuid: 'e1cda789-8c88-53e8-ffff-31c304580c1e',
*   id: 0
* }
*/
function status(iface, cb) {
  if (cb === undefined) {
    cb = iface;
    iface = currentInterface;
  }  
  tools.wpa.status(iface, cb);
}


/**
* @method checkConnection
* @description Returns the state of the network with the specified ssid
* @param {string} ssid Network ssid
* @param {Function(err, result)} Error if unable to get network status, Object with connection details
* {
*   selected: true | false,
*   connected: true | false,
*   ip: 192.168.0.2
* }
*/
function checkConnection(ssid, cb) {
  var result;
  status(function (err, status) {
    if (!err) {
      result = { selected: false, connected: false };
      if (status.hasOwnProperty('ssid') && status.hasOwnProperty('wpa_state')) {
        result.selected = (status.ssid === ssid);
        if (result.selected) result.connected = (status.wpa_state === 'COMPLETED');
        if (result.connected && status.ip) result.ip = status.ip;
      } else {
        err = new Error('Incomplete status object');
      }
    }
    cb(err, result);
  });
}


/**
* @method connectToId
* @description Enables a network, saves the configuration and selects the network with the id provided
* @param {Integer} networkId Network id
* @param {Function(err)} callback Returns error if the process fails
*/
function connectToId(networkId, callback) {
  //console.log('Enabling network ' + networkId + '...');
  tools.wpa.enable_network(currentInterface, networkId, function (err, data) {
    if (err || !data.hasOwnProperty('result') || data.result !== 'OK') return callback(err);
    //console.log('Saving config...');
    tools.wpa.save_config(currentInterface, function (err) {
      if (err || !data.hasOwnProperty('result') || data.result !== 'OK') return callback(err);
      //console.log('Selecting network...');
      tools.wpa.select_network(currentInterface, networkId, function (err, data) {
        if (err || !data.hasOwnProperty('result') || data.result !== 'OK') return callback(err);
        callback();
      });
    });
  });
}


/**
 * @method detectSupplicant
 * @description Looks for a running wpa_supplicant process and if so returns the config file and interface used
 * @param {function} callback (err, iface, configFile) Error if the process failed or no supplicant is running, Interface used, Config file used
 */
function detectSupplicant(callback) {
  exec(commands.detectSupplicant, function (err, stdout) {
    var iface, configFile;
    var result = false;
    if (!err) {
      var lines = stdout.split('\n');
      for (var i = 0; i < lines.length; i++) {
        if (lines[i].indexOf('wlan') > -1) {
          var options = lines[i].split(' ');
          if (options.indexOf('-i') > -1) iface = options[options.indexOf('-i') + 1];
          if (options.indexOf('-c') > -1) configFile = options[options.indexOf('-c') + 1];
          result = true;
        }
      }
    }
    callback(err, iface, configFile);
  });
}


/**
 * @method startSupplicant
 * @description Starts a wpa_supplicant instance
 * @param {object} options Json object where the interface, config file and dns file to use can be specified, otherwhise default values will be selected
 * - iface: Interface to use. Defaults to the currently selected one
 * - config: Configuration file for the supplicant file. Defaults to /etc/wpa_supplicant/wpa_supplicant.conf
 * - dns: DNS file to use. Defaults to /etc/resolv.conf
 * @param {function} callback (err) Error if the process fails
 */
function startSupplicant(options, callback) {
  if (callback === undefined) callback = options; //If no options is passed and just the callback is provided
  
  var iface = options.hasOwnProperty('iface') ? options.iface : currentInterface;
  var configFile = options.hasOwnProperty('config') ? options.config : defaultSupplicantConfigFile;
  var dnsFile = options.hasOwnProperty('dns') ? options.dns : defaultDNSFile;
  exec(replaceInCommand(commands.startSupplicant, { config_file: configFile, interface: iface, dns_file: dnsFile }), callback);
}


/**
 * @method interfaceUp
 * @description Raises the interface provided
 * @param {string} iface Interface to start
 * @param {function} callback (err) Returns an error if the process fails
 */
function interfaceUp(iface, callback) {
  iface = iface || currentInterface;
  exec(replaceInCommand(commands.interfaceUp, { interface: iface }), function (err) {
    callback(err);
  });
}


/**
 * @method interfaceDown
 * @description Drops the interface provided
 * @param {string} iface Interface to stop
 * @param {function} callback (err) Returns an error if the process fails
 */
function interfaceDown(iface, callback) {
  iface = iface || currentInterface;
  exec(replaceInCommand(commands.interfaceDown, { interface: iface }), function (err) {
    callback(err);
  });
}

/**
 * @method restartInterface
 * @description Restarts the interface provided
 * @param {string} iface Interface to restart
 * @param {function} callback (err) Returns an error if the process fails
 */
function restartInterface(iface, callback) {
  async.series([
    async.apply(interfaceDown, iface),
    async.apply(interfaceUp, iface),
  ], callback);
}


/**
 * @method listNetworks
 * @description List the networks in an array, each network has Network ID, SSID, BSSID and FLAGS
 * @param {function} callback (err, networksArray) returns err if the process fails, each network is a Json object that contains network_id, ssid, bssid and flags
 */
function listNetworks(callback) {
  exec(commands.wpaList, function (err, stdout) {
    var tempNetworkJson, parameters, networksArray;

    if (!err) {
      var networksList = stdout.split('\n');
      networksArray = [];
      networksList.splice(0, 2); //Remove headers
      networksList.splice(networksList.length - 1, 1); //Remove footer

      for (var networkIndex in networksList) {
        tempNetworkJson = {};

        parameters = networksList[networkIndex].split('\t');
        tempNetworkJson = {
          network_id: parameters[0],
          ssid: parameters[1],
          bssid: parameters[2],
          flags: parameters[3],
        };
        networksArray.push(tempNetworkJson);
      }
    }

    callback(err, networksArray);
  });
}

/**
 * @method disconnect
 * @description Disconnects from the network on the current interface
 * @param {function} callback (err) returns err if the process fails
 */
function disconnect(callback) {
  exec(commands.wpaDisconnect, callback);
}


/**
 * @method disableSupplicant
 * @description Kills the supplicant process for the specified interface
 * @param {string} iface Interface used by supplicant (If not iface is supplied the current one will be used)
 * @param {function} callback (err) returns err if unable to kill the process
 */
function disableSupplicant(iface, callback) {
  if (callback === undefined) {
    callback = iface; //If no options is passed and just the callback is provided
    iface = currentInterface;
  }
  tools.wpa_supplicant.disable(iface, callback);
}

/**
 * @method replaceInCommand
 * @description Used to replace preset variables strings (e.g.: This is a :VAR text)
 * @param {string} text Text containg the variables to be replaced
 * @param {Object} toReplace Json object that contains the string to find (key) and the replace string (value)
 * @return Returns the text after replacing the variables
 */
function replaceInCommand(text, toReplace) {
  for (var placeHolder in toReplace) {
    text = text.replace(new RegExp(':' + placeHolder.toUpperCase(), 'g'), toReplace[placeHolder]);
  }
  return text;
}

module.exports = {
  check: checkConnection,
  connectTo: connection,
  connectToId: connectToId,
  connect: secureConnection,
  connectOpen: openConnection,
  connectEAP: enterpriseConnection,
  disconnect: disconnect,
  detectSupplicant: detectSupplicant,
  interfaceDown: interfaceDown,
  interfaceUp: interfaceUp,
  killSupplicant: disableSupplicant,
  listNetworks: listNetworks,
  restartInterface: restartInterface,
  scan: scan,
  setCurrentInterface: setCurrentInterface,
  startSupplicant: startSupplicant,
  status: status
}